torch/autograd/saved_variable_default_hooks.py

```diff
@@ -5,3 +5,22 @@ def set_saved_tensors_default_hooks(pack_hook, unpack_hook):
 
 def reset_saved_tensors_default_hooks():
     torch._C._autograd._reset_default_hooks()
+
+def set_save_on_cpu_hooks(pin_memory=False):
```

Comment:
A few questions:
- Should the design of this be a context manager instead of a global flag?
- I remember there are talks about changing hooks to be thread-local: when they are thread-local, does it make more sense for this to be a context manager?
- Is `set_save_on_cpu_hooks` the final API we are exposing to users?

Reply:
A reason why we might want this to be a context manager is, maybe the user only wants to move part of their model's saved variables to CPU

Reply:
Yes ideally, we'd want this to be a a context manager and be thread-local. For now, exposing a context manager could trick the users into thinking that the code is thread-local while it is not. The future API would be:
```python
with torch.autograd.graph.save_on_cpu():
    pass
````

For now, yes, the current API exposed to users would be `set_save_on_cpu_hooks` to set the hooks and `reset_saved_tensors_default_hooks` to clear them. (Working on docstrings right now! could incorporate in this PR if needed)

Reply:
No need to incorporate in this PR, I was curious about the plan.

How soon were you thinking we would change this to be a context manager? If we release this feature as-is in PyTorch 1.10, and then we change this feature to be a context manager in PyTorch 1.11, then we would actually need to do a deprecation cycle and that would not be ideal.

If on the other hand, the plan is to change this to be a context manager before PyTorch 1.10, then everything works out. Maybe this is more of a question for Alban?

Reply:
I'm hoping for an "everything-works-out" scenario but I need to check with Alban that the timelines align.
My PSC is next week and I'm not aiming to have landed the thread-local feature by then. However it would be my first task once I'm done with my PSC tasks. So I expect to start working on it mid-next week?

Reply:
Also, I'm not sure we actually want to depreciate the `set_save_on_cpu_hooks` API.
Similarly to grad mode that has `torch.set_grad_enabled(False)` and the context-manager `with torch.no_grad(): pass`

Reply:
Offering something to modify global state seems like a good way for the user to shoot themselves in the foot. Let's say:
- the user imports a model that they copy-pasted online that sets cpu hooks
- they don't end up running the model and instead run a different model (that didn't globally set cpu hooks)
- Then the user may be confused to find that their new model saves activations to CPU!
- The user might not even notice this -- they might notice a performance drop and be confused.

`torch.set_grad_enabled` is a context manager (https://pytorch.org/docs/stable/generated/torch.set_grad_enabled.html), not global state, so it avoids this problem.

My opinion is that `set_save_on_cpu_hooks` (and other similar APIs we offer directly to the user) should always be context managers, even if they're not thread-local.

Reply:
> torch.set_grad_enabled is a context manager (https://pytorch.org/docs/stable/generated/torch.set_grad_enabled.html), not global state, so it avoids this problem.

Well it is both haha You can use it as a context manager or as a global setter
https://github.com/pytorch/pytorch/blob/093495d3f0985fbe2aea58044798d588c0fe6889/torch/autograd/grad_mode.py#L179

I don't see a reason not to have a functional API for this. It is convenient for debugging purposed in particular or for users that want to change the default across their code.
But I do agree that it is not the best composability-wise and we should not encourage it as the main way to do things.

Reply:
I'm weighing the pros and cons of the two options.
Your example makes a strong point in favor of using a context-manager even when the operation is not thread local.

The con of using a context-manager in that case is it might cause surprising errors to the user.

For instance, this fails (while we are not thread-local):

```python
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.randn(10))

    def forward(self, x):
        with torch.autograd.graph.save_on_cpu():
            y = self.a * (self.a * x)
        return y

net = torch.nn.DataParallel(Model())
x = torch.ones(10, 10)
y = net(x)
```

But this works:

```python
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.randn(10))

    def forward(self, x):
        y = self.a * (self.a * x)
        return y

net = torch.nn.DataParallel(Model())
x = torch.ones(10, 10)
with torch.autograd.graph.save_on_cpu():
    y = net(x)
```

If we were to only keep one API, I'm in favor of the context manager because at least the error will explicitly say "Setting default hooks but they are already defined", while your example shows that leaving an "unclosed" `torch.autograd.graph.set_save_on_cpu_hooks()` can cause sneaky errors.

===

A note regarding `torch.set_grad_enabled`: it's both a context-manager and a function you can call "in the void".
And by the way, I'm not fan of the way it was implemented to allow both usages: the grad mode is set on `__init__` instead of on `__enter__`.

So for instance:

```python
x = torch.ones(5, requires_grad=True)
no_grad_ctx = torch.set_grad_enabled(False)
a = x * x
with no_grad_ctx:
    b = x * x
print(a.requires_grad) # False
print(b.requires_grad) # False
```

Reply:
My bad, I didn't realize `torch.set_grad_enabled` could also be used as a function!
```
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.randn(10))

    def forward(self, x):
        with torch.autograd.graph.save_on_cpu():
            y = self.a * (self.a * x)
        return y

net = torch.nn.DataParallel(Model())
x = torch.ones(10, 10)
y = net(x)
```
What happens with this today?

Reply:
On a machine with only one device, it would work as expected.

On a machine with multiple devices (e.g. more than one GPU), it would raise
`RuntimeError: Setting default hooks but they have already been set. Hint: only one pair of hooks is allowed at a time.`

Reply:
I made the change to context-manager.